### DISCUSSION TITLE
why is streaming audio so difficult?

### DISCUSSION THREAD

**Discussion by tari3x** [roc-streaming/roc-toolkit#336] [Discuss]
I do apologize for the obnoxious title, and I wish there was a "questions" section for this. 

I have tried to use both pulseaudio tcp module and pulseaudio roc to stream audio to my raspberry pi 3, and I still get regular glitches even when I set latency to 500ms. I am utterly perplexed by this. Here are some specific questions:

1. What could possibly be causing this? The bandwidth is over 10mbit/s, that's way more than enough. Ping comes back in under 5ms mostly, and iperf3 claims that my jitter is below 10ms. CPU usage is fine. So where do those glitches with 500ms come from? What am I missing? 
  To compare, I bought a dongle that transmits full HDMI with both video (1920x1680) and sound over 5GHZ wireless, zero glitches. Surely transmitting audio only should be an order of magnitude easier?

2. The quality depends greatly on which app is playing sound on the sender side. mplayer is pretty good, but soundcloud on firefox is a disaster. Does this suggest that the issue is an architectural flaw in pulseaudio and not in raspberry pi?

**Comment by gavv** [comment]
Hi,

Thanks for your interest and for taking time to report this.

First, here are some general ideas of what can cause glitches. Theoretically, every element of the transmission chain can cause them:

* roc on sender
* pulseaudio on sender
* network stack on sender
* network channel
* network stack on receiver
* roc on receiver
* pulseaudio on receiver

What network channel do you use? If you're using WiFi, you can try to run a tool like [WiFi Analyzer](https://play.google.com/store/apps/details?id=com.vrem.wifianalyzer) and see how much other networks (e.g. from neighbors) are there on the same channel as yours. If there are plenty of them, you can try to use another channel. You can also measure the signal strength of your network near the sender and receiver. If it's too low, you can try to move them closer to the access point.

If the channel is noisy, we can try to increase both session latency on receiver (actually it's the jitter buffer size) and FEC block size and redundancy level on sender. Using larger FEC blocks and higher redundancy will eat more traffic, but will also increase the probability of restoring a lost packet. By default we use 20 source packets + 10 repair packets. You can try 20 + 20, 30 + 20, 30 + 30, etc. But you should ensure that the session latency on receiver is higher than FEC block size * packet length on sender.

PulseAudio has troubles with some sound hardware on some boards, including some Raspberry Pi models. So it can sometimes cause glitches even without network streaming. On such boards, I usually switch to ALSA + roc-recv instead of pulseaudio + module-roc-sink-input. You can try to stop and uninstall pulseaudio on raspberry and use bare ALSA.

On some hardware, PulseAudio may work good (without glitches) only on higher latencies. PulseAudio select latency depending on what is requested by currently running apps. Most apps, including media players, usually use the default latency, which is as high as 2s. Real-time apps usually request lower latency and it may cause glitches in PulseAudio, on some hardware. By default module-roc-sink-input and roc-recv request something about 40 ms (it's called "io latency"), but you can try to select a higher value.

Sometimes, glitches are caused by Roc receiver, if the resampler on the receiver is capped by the CPU. What CPU do you have on your Raspberry Pi? We can try to select a lower-quality resampler profile or disable it completely.

Finally, on weak CPUs or loaded systems glitches may be cause by OS scheduler. This can be solved by removing the load, if any, and using real-time scheduling policy for network and audio threads. PulseAudio can automatically do this using rtkit (if it's installed and enabled). Roc command-line tools can be run under chrt tool.

**Comment by gavv** [comment]
Second, here are some more specific suggestions.

We can isolate the problem by running a few tests using `roc-send` and `roc-recv` command-line tools with `-vv` option. Verbose logs can give an idea why there are glitches.

The tests:

1. `roc-send` reading from file, `roc-recv` writing to file (so test only network)

2. `roc-send` reading from file, `roc-recv` writing to PulseAudio (test network + PulseAudio on receiver)

3. `roc-send` reading from PulseAudio, `roc-recv` writing to PulseAudio (test PulseAudio on sender + network + PulseAudio on receiver)

4. `roc-send` reading from file, `roc-recv` writing to ALSA, after temporary disabling PulseAudio on Raspberry (test network + ALSA on receiver)

**Could you run the first test and post the exact commands, their console output, and the recorder file?**

If there are glitches even in the first test, the source of the glitches is the network channel. In this case we can first try to achieve glitch-less playback on the first test, by playing with sender and receiver options. If there are no glitches in the first test, we can continue and try the next tests.

**Comment by gavv** [comment]
> I do apologize for the obnoxious title, and I wish there was a "questions" section for this.

That's OK, we have a discussion label for this. Also, we have a [mailing list](https://roc-project.github.io/roc/docs/about_project/contacts.html), which is more appropriate for extensive discussions like this.

**Comment by gavv** [comment]
Yeah, and answering the title:

> why is streaming audio so difficult?

I think in future versions we will be able to make it easier for the user. Most likely, in 0.4 we will implement adaptive streaming, i.e. Roc will automatically adjust latency and FEC parameters according to the network conditions, so that the user wont need to debug and adjust them manually.

In next releases, maybe in 0.6, I'm planning to spend more time on Roc PulseAudio modules. I hope I might be able to implement the similar technique for I/O latency too, i.e. automatically adjust PulseAudio latency according to how PulseAudio works with given hardware.

Thus, with these two features, Roc will automatically select the best parameters for your network and hardware and make audio streaming a bit easier ;)

Though, I can't provide any estimate. The main limitation is the free time, as usual.

**Comment by tari3x** [comment]
Thanks for the suggestions! I can try them out, but I think that any explanation we come up with needs to account for the fact that mplayer on the sender works almost fine, and soundcloud player doesn't. For instance, is it possible that soundcloud player drip feeds data to pulseaudio, and mplayer feeds a lot more at once? Whether this is possible depends on how the pulseaudio API is structured, which I know nothing about.

I actually noticed underrun messages when playing from soundcloud even when not playing over the network. It's just that in local playback these underruns do not cause any audible glitches. 

I suspect my question "why is streaming audio so difficult?" might have been misunderstood. I am not complaining about the complexity of setting this up from the user perspective. I am wondering about what is difficult about the problem as such from first principles. It seems to me that bandwidths, latencies, and jitters of modern wifi networks are way above what is necessary to transmit lossless sound with very low latency. And yet all the software solutions I tried (pulseaudio-dlna, pulseaudio-tcp, pulseaudio-rtp, pulseaudio-roc) seem to struggle. So I am wondering whether there is something fundamentally difficult about the problem, or whether there is a bug or structural flaw in pulseaudio or elsewhere.

**Comment by gavv** [comment]
> Thanks for the suggestions! I can try them out, but I think that any explanation we come up with needs to account for the fact that mplayer on the sender works almost fine, and soundcloud player doesn't. [...]

I have a habit to debug things step-by-step, but I guess you're right, if there are almost no glitches when streaming from mplayer to raspberry, we can skip these steps.

However, I think it would be useful at least to look at `roc-recv -vv` or `pulseaudio -vvv` logs at the receiver side, when there are glitches. And ideally at `pulseaudio -vvv` logs at the sender side too.

> For instance, is it possible that soundcloud player drip feeds data to pulseaudio, and mplayer feeds a lot more at once?

Maybe.

Could you check what latency is requested by mplayer and soundcloud player when playing to a local sink (without Roc)? To do this, you can run `pactl list sinks` and `pactl list sink-inputs` and see the `Latency` field of appropriate sink-input and sink.

Also, probably it's the fault of the Roc sink. Maybe our sink implementation is somehow incomplete... To check it, we could replace Roc sink with the `roc-recv` tool. You can create a null sink (using `pactl load-module module-null-sink`), run firefox with soundcloud player, connect firefox playback stream to the null sink (via `pavucontrol` GUI), and then run `roc-recv` which reads from null sink's monitor and sends samples to raspberry. When running `roc-recv`, we can try to increase `--io-latency`.

Where are you running the sender? PC?

> I suspect my question "why is streaming audio so difficult?" might have been misunderstood. [...] So I am wondering whether there is something fundamentally difficult about the problem, or whether there is a bug or structural flaw in pulseaudio or elsewhere.

Ah, I see. Well, if you can't get a 250ms latency without glitches on a good WiFi network, that's very strange. And yes, PulseAudio issues are possible.

**Comment by tari3x** [comment]
Got it, I'll set aside some time on Friday to dig into this.

**Comment by gavv** [comment]
Thanks!

**Comment by tari3x** [comment]
So I looked at it again, and to my embarrassment I cannot reproduce the glitches when simply playing music and doing nothing else with network. Not sure what changed since the last time. I guess it's possible some other website in the background was pulling crap. 

However, I still reliably get glitches when trying to load a web page. Is that surprising? When that happens, nothing interesting is printed on the sender side, and the receiver side says

```
D: [alsa-sink-USB Audio] roc_audio: latency monitor: latency=13592 target=8820
D: [alsa-sink-USB Audio] roc_audio: latency monitor: latency=10805 target=8820
D: [alsa-sink-USB Audio] roc_audio: depacketizer: dropping late packet: ts=1194477665 pkt_ts=1194477277
D: [alsa-sink-USB Audio] roc_audio: depacketizer: fetched=1 dropped=1
D: [alsa-sink-USB Audio] roc_audio: watchdog: status: ibD.................
D: [alsa-sink-USB Audio] roc_audio: watchdog: status: ..................i.
D: [alsa-sink-USB Audio] roc_audio: watchdog: status: i....i..............
D: [alsa-sink-USB Audio] roc_audio: depacketizer: dropping late packet: ts=1194526483 pkt_ts=1194526099
D: [alsa-sink-USB Audio] roc_audio: depacketizer: fetched=0 dropped=1
D: [alsa-sink-USB Audio] roc_audio: depacketizer: dropping late packet: ts=1194526924 pkt_ts=1194526408
D: [alsa-sink-USB Audio] roc_audio: depacketizer: fetched=1 dropped=1
D: [alsa-sink-USB Audio] roc_audio: depacketizer: dropping late packet: ts=1194527409 pkt_ts=1194527026
D: [alsa-sink-USB Audio] roc_audio: depacketizer: fetched=1 dropped=1
D: [alsa-sink-USB Audio] roc_audio: depacketizer: dropping late packet: ts=1194528114 pkt_ts=1194527644
D: [alsa-sink-USB Audio] roc_audio: depacketizer: fetched=1 dropped=1
D: [alsa-sink-USB Audio] roc_audio: depacketizer: dropping late packet: ts=1194529040 pkt_ts=1194528571
D: [alsa-sink-USB Audio] roc_audio: depacketizer: fetched=0 dropped=1
D: [alsa-sink-USB Audio] roc_audio: depacketizer: dropping late packet: ts=1194529481 pkt_ts=1194528880
D: [alsa-sink-USB Audio] roc_audio: depacketizer: fetched=1 dropped=1
D: [alsa-sink-USB Audio] roc_audio: watchdog: status: ...........iBIIbIiBD
D: [alsa-sink-USB Audio] roc_audio: depacketizer: dropping late packet: ts=1194532260 pkt_ts=1194531661
D: [alsa-sink-USB Audio] roc_audio: depacketizer: fetched=1 dropped=1
D: [alsa-sink-USB Audio] roc_audio: watchdog: status: ...ibD..............
D: [alsa-sink-USB Audio] roc_audio: latency monitor: latency=2589 target=8820
D: [alsa-sink-USB Audio] roc_audio: depacketizer: ts=1194794654 loss_ratio=0.00067
D: [alsa-sink-USB Audio] roc_audio: latency monitor: latency=13708 target=8820
D: [alsa-sink-USB Audio] roc_audio: latency monitor: latency=13834 target=8820
D: [alsa-sink-USB Audio] roc_audio: latency monitor: latency=13962 target=8820
```

I guess there's no fairness over wifi, so if the page needs to be loaded, it can block all the traffic for a while?

**Comment by gavv** [comment]
Hi, sorry for so late reply.

This log looks like the router delays packet delivery from time to time (the full log would be more helpful though). Apparently, when you're loading a web page, the channel capacity is exhausted and the audio packets are queued until the page is loaded.

This normally not a case, but I saw such situations on some bad networks, e.g. in hotels or at a cafe.

Why it may happen? For example, you channel is very noisy (a lot of overlapping neighbor networks or other source of interference like poorly shielded microwave oven), and the WiFi card has to retry each packet multiple times before it can go to the next packet. Or maybe the router has to switch to a low bandwidth mode depending on some conditions (presence of older devices in the network; using broadcast packets).

What hardware and 802.11 standard are you using? Did you try to measure your WiFi throughput? You can try to install some WiFi analyzer (sniffer) and see how many overlapping networks are there.

As for prioritizing traffic, we have this issue: https://github.com/roc-project/roc/issues/238

**Comment by tari3x** [comment]
Hey, I beat you in slow responding! 

My Wifi throughput is very good, I get 70Mbit down/up, with 2ms ping to 8.8.8.8. My WFH remote connection has almost no perceptible lag. 

I ended up getting a wireless HDMI thing, and it works perfectly, so I'll just stick with that.

**Comment by gavv** [comment]
@tari3x Thanks for sharing!
